### ISSUE TITLE
[BCES00173] SOCOM: Confrontation - Rendering Issues

### ISSUE THREAD

**Issue by ItsLaguna** [RPCS3/rpcs3#14674]
### Quick summary

Game has issues while rendering maps, sometimes structures and models disappear and then reappear after a few seconds.
What seems to be shadows are also bugged.

The game requires Precise ZCull to alleviate the vast majority of the rendering issues, if set to Approximate, more than half of the map being played barely renders around the player.

<details>
<summary>Precise ZCull</summary>

https://github.com/user-attachments/assets/2aea36cc-53d4-42b4-9594-a896ae9f1698
</details>

<details>
<summary>Approximate ZCull</summary>

https://github.com/user-attachments/assets/a873bf08-3dfb-4da7-b172-9fd402de76c2
</details>

Also a few examples of water and shadows bugging:
<details>
<summary>Videos</summary>
<br>

https://github.com/RPCS3/rpcs3/assets/108736840/b5ea4dbf-972f-4545-b8b5-3f1a9068c57e

https://github.com/RPCS3/rpcs3/assets/108736840/8a775767-866d-4a62-bec6-040d78b4b281

https://github.com/RPCS3/rpcs3/assets/108736840/70ac01d9-3991-46ba-ab17-929c58485807

https://github.com/RPCS3/rpcs3/assets/108736840/b37e8884-278a-4669-804c-fdeef409c3c3

</details>

I saw a "similar" issue with MGS4 where it needed Accurate xfloat to prevent specific sections in an Act from disappearing at specific angles, but that didn't fix it for this game... in fact, it bugs the color of some textures.

None of the RSX buffers seem to affect this either.

### Attach a log file

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/12709631/RPCS3.log.gz)

### Attach capture files for visual issues

I'm not completely sure how RSX Captures work, but I tried to capture the exact frame of some issues, for now there's just one for the water and another of a small part of the map disappearing. If it's not enough I'll try to upload more.

[Captures](https://www.mediafire.com/file/a79ejzvra0h8uh7/captures.7z/file)
[New captures](https://www.mediafire.com/file/wtm9ydlgbeekdm2/new+captures.7z/file)

<details>
<summary>Videos of each if needed</summary>
<br>

https://github.com/RPCS3/rpcs3/assets/108736840/09e9274c-ab30-4551-86bb-9cf8409cd004

https://github.com/RPCS3/rpcs3/assets/108736840/9f834105-c05d-4eeb-9508-8ad39276682f

https://github.com/RPCS3/rpcs3/assets/108736840/0cb96885-82c6-4b5f-ae4c-0b00f7416323

https://github.com/RPCS3/rpcs3/assets/108736840/52b6726a-7e7a-4fd1-83f4-7cb8f54af07d

</details>

### System configuration

- OS: Windows 10 Pro
- CPU: Ryzen 7 5800X
- RAM: 32 GB
- GPU: Nvidia RTX 2060
- Driver version: 536.67.0.0

**Comment by Greediez** [comment]
Same issues as well for me….i7 9700F CPU… RTX 2070 Super

**Comment by SIlEnT-KiLLeRc** [comment]
yeah i have the same issues. i have 100 hours in testing configs with no luck.
cpu 13600k
ram 32gb 4800
Nividia 4070ti
always latest drivers installed.

**Comment by Nursacoma** [comment]
same issue and reported here : #14231

**Comment by shankly1985** [comment]
I am experiencing the same issue.

**Comment by ItsLaguna** [comment]
Sadly still an issue.

**Comment by sebringers** [comment]
Users are still experiencing issues. since these are really old and overdue is it possible to receive a status update on this?

**Comment by ItsLaguna** [comment]
Updated the main comment as the rsx captures had been taken down.

**Comment by MarkonScon** [comment]
Still experiencing this issue

**Comment by digant73** [comment]
please re-test on latest build. Hopefully some textures and reflections issues have been fixed by recent build 0.0.39-18790

**Comment by joinaunion** [comment]
@kd-11 Is this correct capture you need? I've zipped it as requested.
https://mega.nz/file/6w5xlCiD#luy7pnOy0D48QwcQM6hsJ29N-K2usOLFk70NySJ7sOU

<img width="339" height="357" alt="Image" src="https://github.com/user-attachments/assets/3e14c6d8-2747-41db-97b8-e60edf1648b8" />

**Comment by joinaunion** [comment]
Hi @kd-11,

Following up on the SOCOM Confrontation visibility/pop-in issue (objects only appearing when extremely close).

I replayed the full capture you already have (`gfxrecon_capture_20251116T160925.gfxr`) with Vulkan Validation Layers enabled. The resulting log is small (24 KB) and shows several consistent validation errors that strongly correlate with the culling/pop-in problem affecting many users.

### Key Validation Findings:

- **Occlusion Query / Render Pass Mismatches** (most relevant to the bug):  
  Repeated `VUID-vkCmdEndQuery-None-07007`: `vkCmdEndQuery()` called for queries started outside a render pass.  
  Repeated `VUID-vkCmdEndRenderPass-None-07004`: Queries begun in a subpass but never properly ended inside the same subpass.  
  These appear across many frames (examples: 195, 541, 697, 1027+, 1678, 1828, 1892, etc.).

  → This likely disrupts RPCS3’s occlusion query implementation for visibility culling/ZCull, causing most distant and mid-range objects to be incorrectly culled.

- **Fragment Shader Output Issues**:  
  Repeated `Undefined-Value-ShaderOutputNotConsumed` warnings (starting at Frame 19):  
  Fragment shaders write to output locations 0–3, but the render pass has no matching `pColorAttachments`.  
  This can make materials render invisible or broken even if they pass culling.

- **Other**:  
  `VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT` used without the `bufferDeviceAddress` feature enabled during replay.

The replay completes successfully and the game is visible, but these validation hits point to problems in RSX → Vulkan translation, particularly around query handling and render pass management.

I’ve attached the validation log: **socom_deep_validation.log** (24 KB).

Let me know if this helps or if you’d like me to test specific flags, provide a trimmed capture, or run any other diagnostics.

Thanks!

[socom_deep_validation.log](https://github.com/user-attachments/files/26335496/socom_deep_validation.log)

**Comment by kd-11** [comment]
None of the findings are real issues:
1. Occlusion query - This is an intentional speedhack. It is enabled per-driver where we know it doesn't cause issues. You can enable "strict rendering mode" to disable the hack and follow vulkan guidelines at a cost to performance.
2. FP outputs - The bindings are the source of truth. We just find a shader that works and bind it, unused outputs are fine if it outputs depth. Otherwise we'd have millions of shader variants.
3. We don't use BDA, so probably just some VMA noise. Nothing to worry about.

**Comment by joinaunion** [comment]
Hi @kd-11,
Thanks for the explanation. I already tested with Strict Rendering Mode enabled. But unfortunately it did not reduce the object pop-in / missing geometry issue.
The repeated occlusion query errors (vkCmdEndQuery and vkCmdEndRenderPass violations) in the validation log still appear to be the main cause behind the widespread culling problems that many users are experiencing with SOCOM Confrontation.
Is there anything else worth trying from the user side, or is this primarily something that needs improvement on the emulation side?Thanks again for looking into it.

**Comment by sebringers** [comment]
I cleared my configuration and restored RPCS3 defaults. using the link to the Dropbox below. I included a clip from multiple times, the session log with a crash at the end, and my current configuration. There are minor pop-ins during the session, but it worsens significantly around 2:40 in the capture. Overall, performance has improved since my last test around November 2025.

link to the log 
https://www.dropbox.com/scl/fi/ru8fl0he6fd81tfgd589a/log.zip?rlkey=j5p6fovy3hrt58glzrh809f2d&st=878ditsg&dl=0

**Comment by AniLeo** [comment]
File a compatibility report for that in the forums, compatibility is not tracked through GitHub

**Comment by kd-11** [comment]
> Thanks for the explanation. I already tested with Strict Rendering Mode enabled. But unfortunately it did not reduce the object pop-in / missing geometry issue.
The repeated occlusion query errors (vkCmdEndQuery and vkCmdEndRenderPass violations) in the validation log still appear to be the main cause behind the widespread culling problems that many users are experiencing with SOCOM Confrontation.

Strict rendering mode completely resolves that issue so if SRM is not helping, then that is not the issue.

> Is there anything else worth trying from the user side, or is this primarily something that needs improvement on the emulation side?Thanks again for looking into it.

Only host GPU labels could do anything here but the feature is kinda busted at the moment. I'm fairly certain I know why it has occlusion holes and it's really not an occlusion problem just a data problem on the game's side. They're not waiting for the occlusions to become available asynchronously, they're inferring readiness based on some other event. You can also try the OpenGL backend which does emits inline (though we don't force flush/finish anyway) and maybe your GPU is fast enough to make the flickering less obvious.

**Comment by sebringers** [comment]
Do we need to try a new rsx cap or is it not worth it

**Comment by sebringers** [comment]
I tried to get an RSX capture of a match in-game, but it kept crashing, and it didn't write to the folder. I did get one of the main menu. I also have the log from that same capture.

[RPCS3.log](https://github.com/user-attachments/files/27260328/RPCS3.log)

[BCUS98183_20260430161531_capture.rrc.gz](https://github.com/user-attachments/files/27260319/BCUS98183_20260430161531_capture.rrc.gz)

I also have stub ppu traps set to 1 because it was recommended by the log/discord bot. Im not sure if you want that disabled. I'm not sure why the capture shows the player models blacked out because it doesn't look like that in-game could just be the way the RSX cap works.

**Comment by Ashmund** [comment]
Just wanted to note that I have proven the issue and fix for this rendering issue:


Issue: 

GCM_SET_USER_COMMAND (0xEB00) woke the PPU handler as soon as the FIFO method was decoded. Host occlusion queries are async (Vulkan may still have the query on the current command buffer).

ZCULL_control::sync() also no-op’d when m_critical_reports_in_flight == 0. That counter only moves after a report page is accessed and MAIN/LOCAL shortcuts turn off. While shortcuts are still on, even an explicit sync would not retire vis reports.

So stage 1’s interrupt could run — and consume could run — before the GPU had published begin/end. ICE then saw -1 (keep stale vis) or a published 0 (hide). Distant in-view geometry stayed gone until a later frame happened to sample correctly. That is the pop-in.

Proven fix:

Call ZCULL sync with a FIFO barrier.
That path must not early-out on m_critical_reports_in_flight == 0.
It must not take the relaxed-ZCULL shortcut.
Pending ZPASS reports retire for real (get_occlusion_query_result). If the query is still current on Vulkan, force a command buffer submission to retire the results.

@kd-11

**Comment by kd-11** [comment]
@Ashmund 
This proposal is a hackfix focused on making the symptoms go away, not checking the underlying cause. The whole async explanation is a bit of a misdirection.
GCM_SET_USER_COMMAND is not an access ordering command at all. There is no guarantee that firing the callback interrupt also means all rendering operations before this point have also completed. You do actually need an ordering command for that like the various XXX_semaphore_release ones. Many things are also async on real RSX afterall.
In general RSX runs frontend PFIFO and backend PGRAPH in parallel, so maybe SET_REPORT is frontend-only? I find it hard to believe since it is accessing registers filled in by the backend. That really needs some detailed hardware tests to prove. If it is, then things change and we need to rethink the synchronization model a bit more.

EDIT: Or maybe it's just a game bug and they forgot to add the semaphore but it just works on PS3 because performance on every PS3 behaves the same way. In that case the same solution would be required for games that have the same issue. Definitely something I need to look into a bit more.

**Comment by elad335** [comment]
There may be some delay in real PS3 for GCM_SET_USER_COMMAND, RPCS3 handles such things incredibly fast - too fast.

**Comment by kd-11** [comment]
Yea, I plan to measure the overhead. Vulkan especially just tramples through the FIFO queue at a really fast pace without ever stopping since its a recorded API, so you're basically seeing what would have taken 30ms on a PS3 finish on PC in less than 1ms.

**Comment by Ashmund** [comment]
> Yea, I plan to measure the overhead. Vulkan especially just tramples through the FIFO queue at a really fast pace without ever stopping since its a recorded API, so you're basically seeing what would have taken 30ms on a PS3 finish on PC in less than 1ms.

Hey kd, I'm actually part of the main Confrontation community that plays on emulator, I don't play the game much but I am staff of the server they play on so wanted to fix this as it's been plaguing Confrontation for a long time.

I'm glad to hear this is at least moving this forward as I know it's been a pain in the ass for everyone, I can always have them test this fix over a week or so and report any issues from it.

A side effect to this fix is that things in-game that never rendered properly now do, apart from the pop-in objects/walls, etc.. the water in the game was also flickering/creating artifacts, I ran around the water, shot it, went in and out of it and it was 1:1 to what the water acted like normally on PS3, on top of this the game also feels "smoother" too though I don't know how to meaningfully judge that.

**Comment by kd-11** [comment]
> I'm glad to hear this is at least moving this forward as I know it's been a pain in the ass for everyone, I can always have them test this fix over a week or so and report any issues from it.

Well, the fix is definitely not getting merged/upstreamed due to the way it is implemented, but you can ofc use the build within the community while we figure out why it behaves the way it does.

The problem with titles like these and why they see little direct testing from devs is the online-only requirement. It just makes it impossible for devs to jump in and out with savestates and capture tools to measure things in an apples-to-apples manner from build to build. If you have some way to load maps without connecting to a match it would really help, but I'm not hopeful that is a possibility at this time.

**Comment by Ashmund** [comment]
> > I'm glad to hear this is at least moving this forward as I know it's been a pain in the ass for everyone, I can always have them test this fix over a week or so and report any issues from it.
> 
> Well, the fix is definitely not getting merged/upstreamed due to the way it is implemented, but you can ofc use the build within the community while we figure out why it behaves the way it does.
> 
> The problem with titles like these and why they see little direct testing from devs is the online-only requirement. It just makes it impossible for devs to jump in and out with savestates and capture tools to measure things in an apples-to-apples manner from build to build. If you have some way to load maps without connecting to a match it would really help, but I'm not hopeful that is a possibility at this time.

I found a method for this in the game but it is currently disabled, not sure if it was part of debug that wasn't shipped in the ELF or what but I will keep investigating, I have passed the RPCS3.exe with the edits I did to the community and so far nothing but good news, rendering solved, water solved, they can play without any issues at the moment, so even if this isn't the root of the issue I'm glad we've found something as a temporary fix so these guys can enjoy their game.

Also I wanted to note that I definitely didn't expect this to get merged into RPCS3 at any point, I'm more so just doing work on figuring out what it could be and reporting it to you so hopefully something I'm doing gives you a good path to follow and find the real cause.

**Comment by kd-11** [comment]
> I found a method for this in the game but it is currently disabled, not sure if it was part of debug that wasn't shipped in the ELF or what but I will keep investigating.

Thanks. If you find anything let me know.

**Comment by Ashmund** [comment]
> > I found a method for this in the game but it is currently disabled, not sure if it was part of debug that wasn't shipped in the ELF or what but I will keep investigating.
> 
> Thanks. If you find anything let me know.

Hey kd, just wanted to let you know that I wasn't able to find anything of value in terms of a way the game will let you load into an offline map, so I decided to make my own, so far we are able to load directly into a map from the start menu, completely bypassing online, it will obviously be stripped of many of the online things you'd expect (round time, round count, scoreboard, etc..) but all other gameplay features will be working within the next few days, this is being done via a .yml patch and a .py file run via powershell that is forcing the game to change states on request.

I'll keep you updated on the progress, once this is done what is the best way to get this to you? I personally would feel the most comfortable sharing this to you privately if possible, I don't know if Github has some sort of PM feature or if there's a way to do so here.

This is the current state of the offline loader.

https://streamable.com/zd5qz7 (not sure of a better way to upload it atm)

**Comment by Ashmund** [comment]
I've spent some time fully understanding how Confrontation's rendering works, here's what I know that I believe is relevant:

Confrontation registers cellGcmSetUserHandler. The callback is 0x32CC80. It does not read GET_REPORT memory. It stores the low 8 bits of the cause into a TOC table:

0032cc80  rldicl  r9, r3, 0x38, 0x3f   ; slot = (cause >> 8) & 1
0032cc84  lwz     r0, 0x1c14(r2)       ; status table
0032cc88  clrlwi  r3, r3, 0x18         ; stage = cause & 0xff
0032cc98  stw     r3, 4(r9)
0032cc9c  blr

The same frame path issues occlusion reports, then immediately emits GCM_SET_USER_COMMAND:

0039bc44  bl      0x545718             ; OcclusionManager GET_REPORT issue
0039bc5c  slwi    r4, r4, 8
0039bc60  ori     r4, r4, 1            ; cause = (frame_slot << 8) | stage 1
0039bc68  bl      0x5db1f0
0x5DB1F0 writes a 4-word GCM packet, including 0x0004EB00 (NV4097_SET_USER_COMMAND, count 1) and the cause. Stages 2 / 3 / 5 are the same helper with | 2, | 3, | 5.

After stage 3 it busy-waits on that status word, not on the report timestamp:

0039c47c  lwz     r0, 4(r31)           ; status_table[slot].stage
0039c484  cmpwi   cr7, r0, 0
0039c488  bne     cr7, 0x39c4fc        ; proceed once handler stored a non-zero stage
… timebase spin, then:
0039c4f0  ble     cr7, 0x39c47c
0039c560  bl      0x582978             ; consume
Consume (0x582B0C) is:

Load report end at +0x18. If end == -1, skip (keep last vis bit).
delta = end - begin (begin at +0x08).
Unsigned cmplwi cr7, delta, 19. delta > 19 → vis bit 1; delta ≤ 19 → vis bit 0.
00582b10  lwz     r11, 0x18(r9)
00582b14  cmpwi   cr7, r11, -1
00582b18  beq     cr7, 0x582a04
00582b1c  lwz     r0, 8(r9)
00582b24  subf    r0, r0, r11
00582b28  cmplwi  cr7, r0, 0x13
00582b30  bgt     cr7, 0x582b38
00582b34  li      r9, 0

So the game’s readiness signal for this path is “USER_COMMAND handler ran.” It will hide on a published delta ≤ 19. It will not hide on still-unwritten -1 (it keeps the previous bit).

@kd-11 

Hope this can shed some light on the issue.

**Comment by kd-11** [comment]
Thanks. While it doesn't confirm hardware behavior it does at least tell us that the devs expect GCM_SET_USER_COMMAND to fire when all previous work has been completed. The planned hardware tests will prove this.

**Comment by kd-11** [comment]
PR: https://github.com/RPCS3/rpcs3/pull/19549

The hardware test is attached below:

[user_command.zip](https://github.com/user-attachments/files/32479901/user_command.zip)

The test showed that USER_COMMAND indeed completely waits until everything has completed and the RSX frontend and backend are ready. Also a related test proved that our way of calling the callback handlers was broken. Not directly related to this issue, but maybe it helps a different game.
